src/libraries/Common/src/Interop/Linux/OpenLdap/Interop.Ldap.cs

```diff
@@ -67,6 +68,26 @@ internal static partial class Ldap
     {
         static Ldap()
         {
+            // Register callback that tries to load other libraries when the default library "libldap-2.4.so.2" not found
+            AssemblyLoadContext.Default.ResolvingUnmanagedDll += (assembly, ldapName) =>
```

Comment:
Looking at [the doc](https://learn.microsoft.com/dotnet/core/dependency-loading/loading-unmanaged), there might be a better mechanism for this. I'm wondering if we should consider [`NativeLibrary.SetDllImportResolver(Assembly, DllImportResolver)`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary.setdllimportresolver#system-runtime-interopservices-nativelibrary-setdllimportresolver(system-reflection-assembly-system-runtime-interopservices-dllimportresolver)). @elinor-fung Do you have guidance on which of the resolving mechansims we should be using in this case? I think the `DllImportResolver` callback is always called so perhaps the approach here is correct, but want to verify.

The gist of the scenario is _most_ of the time the unmanaged binary is present on the machine. However, when it isn't, we would like to try newer versions in the case the system has been updated.

Reply:
`DllImportResolver` gets called first - before even the 'normal' attempt to load. `ResolvingUnmanagedDll` is fired last - only if all other resolution has failed.

Considerations for choosing between them:
1. Should your handler take precedence over other resolution methods or just be a fallback?
    - `DllImportResolver` will be called before the default attempt to load `Libraries.OpenLdap`. You can fully define the order of what you attempt to load (so you can say use `libldap-2.5.so.0` as long as it is found - regardless of whether `libldap-2.4.so.2` is present)
      - If your assembly has a lot of other p/invokes that your handler doesn't care about, this may also mean a lot of extra/unnecessary calls to a handler
    - `ResolvingUnmanagedDll` is only called as a fallback. It will not be called when other forms of resolution succeed.
2. Will there be multiple such handlers?
    - Only one `DllImportResolver` is allowed per `Assembly`. If one is already set for that assembly, `SetDllImportResolver` will fail.
    - `ResolvingUnmanagedDll` is an event, so you can have as many handlers as you want

> The gist of the scenario is most of the time the unmanaged binary is present on the machine. However, when it isn't, we would like to try newer versions in the case the system has been updated.

I think this would indicate only wanting the fallback for (1) - `ResolvingUnmanagedDll`.

Perhaps a larger concern is (2) - because this file is under Common and it has no guarantee that all the assemblies using it do not set a `DllImportResolver`.

In terms of setting a precedent in `libraries/Common/**/Interop/` files for handling this kind of thing, I think `ResolvingUnmanagedDll` would be the better option.

Reply:
> If your assembly has a lot of other p/invokes that your handler doesn't care about, this may also mean a lot of extra/unnecessary calls to a handler

Yes, we have lots of other p/invokes, mostly those are different for each platform, I wouldn't like to handle them all in the handler 

> ResolvingUnmanagedDll is only called as a fallback. It will not be called when other forms of resolution succeed.

I think that is what I need, the issue is only happening for one Linux library, I want only handle that failure, if it happens, don't want to change/touch other libraries on other platforms

> I think this would indicate only wanting the fallback for (1) - ResolvingUnmanagedDll.

Thank you for broad explanation, completely agree